pkg/controllers/management/kontainerdrivermetadata/data.go

```diff
@@ -387,97 +400,252 @@ func getWindowsName(str string) string {
+	if err != nil {
+		return nil, err
+	}
+
 	k8sVersionRKESystemImages := map[string]interface{}{}
 	k8sVersionSvcOptions := map[string]v3.KubernetesServicesOptions{}
 
 	for majorVersion, k8sVersion := range maxVersionForMajorK8sVersion {
 		if !deprecated[k8sVersion] {
 			k8sVersionRKESystemImages[k8sVersion] = nil
-			k8sVersionSvcOptions[k8sVersion] = K8sVersionServiceOptions[majorVersion]
+			k8sVersionSvcOptions[k8sVersion] = k8sVersionServiceOptions[majorVersion]
 		}
 	}
 
-	var keys []string
-	for k := range maxVersionForMajorK8sVersion {
-		if !deprecated[k] {
-			keys = append(keys, k)
-		}
+	k8sCurrRKEdata, err := marshal(k8sVersionRKESystemImages)
+	if err != nil {
+		return nil, err
+	}
+
+	k8sSvcOptionData, err := marshal(k8sVersionSvcOptions)
+	if err != nil {
+		return nil, err
 	}
-	sort.Strings(keys)
 
-	return SaveSettings(k8sVersionRKESystemImages, k8sVersionSvcOptions, DefaultK8sVersions, rancherVersion, keys, deprecated)
+	deprecatedData, err := marshal(deprecated)
+	if err != nil {
+		return nil, err
+	}
+
+	minVersion := maxVersions[0]
+	maxVersion := util.GetTagMajorVersion(defaultK8sVersion)
+	uiSupported := fmt.Sprintf(">=%s.x <=%s.x", minVersion, maxVersion)
+	uiDefaultRange := fmt.Sprintf("<=%s.x", maxVersion)
+
+	return map[string]string{
+		"k8s-versions-current":            strings.Join(k8sVersionsCurrent, ","),
+		"k8s-version":                     defaultK8sVersion,
+		"k8s-versions-deprecated":         deprecatedData,
+		"ui-k8s-default-version-range":    uiDefaultRange,
+		"ui-k8s-supported-versions-range": uiSupported,
+		"k8s-version-to-images":           k8sCurrRKEdata,
+		"k8s-version-to-service-options":  k8sSvcOptionData,
+	}, nil
 }
 
-func SaveSettings(k8sCurrVersions map[string]interface{},
-	k8sVersionSvcOptions map[string]v3.KubernetesServicesOptions, rancherDefaultK8sVersions map[string]string,
-	rancherVersion string, maxVersions []string, deprecated map[string]bool) error {
+func (md *MetadataController) updateSettingFromFields(updateField map[string]string, defaultValue bool) error {
+	if defaultValue {
+		for key := range rancherUpdateSettingMap {
+			val, ok := updateField[key]
+			if !ok {
+				return fmt.Errorf("driverMetadata: default value not present for setting %s", key)
+			}
+			setting, err := md.SettingLister.Get("", key)
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					return err
+				}
+				setting, err = md.Settings.Get(key, metav1.GetOptions{})
+				if err != nil {
+					return err
+				}
+			}
+			settingCopy := setting.DeepCopy()
+			settingCopy.Default = val
 
-	k8sCurrVersionData, err := marshal(k8sCurrVersions)
-	if err != nil {
-		return err
+			_, err = md.Settings.Update(settingCopy)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
 	}
-	var versions []string
-	for k := range k8sCurrVersions {
-		versions = append(versions, k)
+
+	// normal setting update
+	for key, setting := range rancherUpdateSettingMap {
```

Comment:
Maybe we should skip updating the settings updated by the user - by using the labels